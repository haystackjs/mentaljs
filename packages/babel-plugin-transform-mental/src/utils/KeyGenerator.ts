import * as UUID from 'uuid';

export interface KeyGenerator {
    generateKey(): string;
}

export const KeyGeneratorDefault: KeyGenerator = {
    generateKey() {
        return UUID.v4();
    }
}