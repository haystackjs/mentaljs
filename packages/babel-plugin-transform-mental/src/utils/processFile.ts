import traverse from '@babel/traverse';
import { createTraversal } from './createTraversal';
import * as babelParser from '@babel/parser';
import babelGenerator from '@babel/generator';
import { Node } from '@babel/types';
import { KeyGenerator, KeyGeneratorDefault } from './KeyGenerator';

function parse(src: string) {
    return babelParser.parse(src, {
        plugins: [
            'asyncGenerators',
            'classProperties',
            'dynamicImport',
            'functionBind',
            'jsx',
            'objectRestSpread',
            'typescript'
        ],
        sourceType: 'module'
    });
}

function generate(ast: Node) {
    return babelGenerator(ast, {
        quotes: 'single'
    }).code;
}

export function processFile(content: string, keyGenerator: KeyGenerator = KeyGeneratorDefault) {
    let res = parse(content);
    traverse(res, createTraversal(keyGenerator));
    return generate(res);
}