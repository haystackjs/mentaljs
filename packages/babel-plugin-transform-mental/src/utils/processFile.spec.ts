import { processFile } from "./processFile";
import { KeyGenerator } from "./KeyGenerator";

class TestKeyGenerator implements KeyGenerator {
    private values: string[];
    private index = 0;
    constructor(values: string[]) {
        this.values = values;
    }
    generateKey() {
        return this.values[this.index++ % this.values.length];
    }
}

describe('babel', () => {
    it('should extract static styles', () => {
        let source = `
            export class Bar extends React.Component {
                render() {
                    let a = 1;
                    return (
                        <XView margin={20} marginTop={a}>
                            Hello!
                        </XView>
                    );
                }
            }
        `;
        expect(processFile(source, new TestKeyGenerator(['key']))).toMatchSnapshot();
    });

    it('should replace with static div for static-only views', () => {
        let source = `
            export class Bar extends React.Component {
                render() {
                    return (
                        <XView margin={40}>
                            Hello!
                        </XView>
                    );
                }
            }
        `;
        expect(processFile(source, new TestKeyGenerator(['key']))).toMatchSnapshot();
    });
});