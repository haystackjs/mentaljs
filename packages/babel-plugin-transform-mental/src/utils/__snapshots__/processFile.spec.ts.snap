// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel should extract static styles 1`] = `
"import { calculateStyles } from \\"openland-x-styles/calculateStyles\\";

var ___style_key_n = calculateStyles({
  margin: 20
});

export class Bar extends React.Component {
  render() {
    let a = 1;
    return <XView marginTop={a} __styleClassName={___style_key_n}>
                            Hello!
                        </XView>;
  }

}"
`;

exports[`babel should replace with static div for static-only views 1`] = `
"import { calculateStyles } from \\"openland-x-styles/calculateStyles\\";

var ___style_key_n = calculateStyles({
  margin: 20
});

export class Bar extends React.Component {
  render() {
    return <div className={___style_key_n}>
                            Hello!
                        </div>;
  }

}"
`;
